lib/dal/src/workspace_snapshot/split_snapshot.rs

```diff
@@ -0,0 +1,1634 @@
+            inferred_connection_graph: Arc::new(RwLock::new(None)),
+        })
+    }
+
+    pub async fn find_for_change_set(
+        ctx: &DalContext,
+        change_set_id: ChangeSetId,
+    ) -> WorkspaceSnapshotResult<Self> {
+        // There's a race between finding which address to retrieve and actually retrieving it
+        // where it's possible for the content at the address to be garbage collected, and no
+        // longer be retrievable. We'll re-fetch which snapshot address to use, and will retry,
+        // hoping we don't get unlucky every time
+        let mut retries: u8 = 5;
+
+        while retries > 0 {
+            retries -= 1;
+
+            let row = ctx
+                .txns()
+                .await?
+                .pg()
+                .query_opt(
+                    "SELECT workspace_snapshot_address FROM change_set_pointers WHERE id = $1",
+                    &[&change_set_id],
+                )
+                .await?
+                .ok_or(
+                    WorkspaceSnapshotError::ChangeSetMissingWorkspaceSnapshotAddress(change_set_id),
+                )?;
+
+            let address: WorkspaceSnapshotAddress = row.try_get("workspace_snapshot_address")?;
+
+            match Self::find(ctx, address).await {
+                Ok(snapshot) => return Ok(snapshot),
+                Err(
+                    WorkspaceSnapshotError::SplitSnapshotSuperGraphMissingAtAddress(_)
+                    | WorkspaceSnapshotError::SplitSnapshotSubGraphMissingAtAddress(_),
+                ) => {
+                    warn!(
+                        "Unable to retrieve split snapshot {:?} for change set {:?}. Retries remaining: {}",
+                        address, change_set_id, retries
+                    );
+                    tokio::time::sleep(tokio::time::Duration::from_millis(5)).await;
+                    continue;
+                }
+                Err(e) => return Err(e),
+            }
+        }
+
+        error!(
+            "Retries exceeded trying to fetch split snapshot for change set {:?}",
+            change_set_id
+        );
+
+        Err(WorkspaceSnapshotError::WorkspaceSnapshotNotFetched)
+    }
+
+    pub async fn write(
+        &self,
+        ctx: &DalContext,
+    ) -> WorkspaceSnapshotResult<WorkspaceSnapshotAddress> {
+        let self_clone = self.clone();
+        let layer_db = ctx.layer_db().clone();
+        let events_tenancy = ctx.events_tenancy();
+        let events_actor = ctx.events_actor();
+
+        let supergraph_address = slow_rt::spawn(async move {
+            let mut working_copy = self_clone.working_copy_mut().await;
+            let start = Instant::now();
+            working_copy.cleanup_and_merkle_tree_hash();
+            warn!("cleaned up working copy in {:?}", start.elapsed());
+
+            warn!(
+                "current addresses: {:?}",
+                working_copy.supergraph().addresses()
+            );
```

Comment:
Could this become noisy quickly?

Reply:
yes for sure we'll need to tone down the warns here, to debug